pkg/labels/labels.go

```diff
@@ -186,59 +203,34 @@ func (l *Label) Matches(target *Label) bool {
 	return l.IsAllLabel() || l.Equals(target)
 }
 
-func (l *Label) Covers(path string) bool {
-	key := l.AbsoluteKey()
-
-	// Step 1: Path of node must be a prefix of the label key
-	if strings.HasPrefix(key, path) {
-		// Step 2: Coverage is only met on either a full match or if the prefix covers an entire layer
-		if len(key) == len(path) || key[len(path)] == '.' {
-			return true
-		}
-	}
-
-	return false
-}
-
-// Interface to implement in order to attach a label to it
-type LabelAttachment interface {
-	GetLabelParent() LabelAttachment
-	Path() string
-}
-
 // Resolve resolves the absolute key path for this Label from policyNode.
-func (l *Label) Resolve(policyNode LabelAttachment) {
-	// FIXME: the HasPrefix should be using daemon.config.ValidLabelPrefixes
-	if l.Source != common.ReservedLabelSource &&
-		!strings.HasPrefix(l.Key, common.GlobalLabelPrefix) &&
-		!strings.HasPrefix(l.Key, common.K8sPodNamespaceLabel) {
-		k := l.Key
-		node := policyNode
-
-		for strings.HasPrefix(k, "../") {
-			k = k[3:]
-			node = node.GetLabelParent()
-			if node == nil {
-				log.Warningf("Could not resolve label %+v, reached root\n", l)
-				return
-			}
-		}
+func (l *Label) Resolve(owner LabelOwner) {
+	l.SetOwner(owner)
 
-		l.absKey = node.Path() + "." + k
-	} else {
-		l.absKey = l.Key
-	}
+	// Force generation of absolute key
+	l.AbsoluteKey()
 
-	log.Debugf("Resolved label %s to %s\n", l.String(), l.absKey)
+	log.Debugf("Resolved label %s to path %s\n", l.String(), l.absKey)
 }
 
 // AbsoluteKey if set returns the absolute key path, otherwise returns the label's Key.
 func (l *Label) AbsoluteKey() string {
-	if l.absKey != "" {
-		return l.absKey
+	if l.absKey == "" {
+		// Never translate using an owner if a reserved label
+		// FIXME: the HasPrefix should be using daemon.config.ValidLabelPrefixes
```

Comment:
With this change. Only the `root.` prefix is a valid prefix. I will remove this FIXME.

What we can talk about though is validating absolute label names as they are loaded as policy and then reject them. This way it will become trivial to notice if a policy contains elements which are not covered by the configured prefix.